### PR TITLE
EPL-744: aws step function support

### PR DESCRIPTION
**Pull request by shawnrusaw-wf**
@kevinfrommelt-wf @ericlarssen-wf @andrewallen-wf 

Straightforward addition of AWS Step Function support. This will give us a UI for no additional development effort.

Key points:

1. script to convert existing .yaml machines into .json specifications as per https://states-language.net/spec.html
2. the same lambda function runs at every state, taking advantage of the built in routing in the framework
3. all retries etc. are now handled by AWS Step Functions infrastructure. these settings can easily be tweaked as per https://states-language.net/spec.html
4. AWS Step Functions support wait states, so no longer any need to use SQS

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Raven
Number of Findings: 0

**Comment by shawnrusaw-wf** [comment]
screen shot 2016-12-03 at 7 15 59 am

**Comment by shawnrusaw-wf** [comment]
screen shot 2016-12-03 at 7 16 41 am

**Comment by beaulyddon-wf** [comment]
Looks great

**Comment by shawnrusaw-wf** [comment]
@kevinfrommelt-wf @ericlarssen-wf any chance you guys can get to this ?

File: aws_lambda_fsm/handler.py [hunk 1]
```diff
@@ -56,6 +56,31 @@ def _process_payload(payload_str, obj):
     fsm.dispatch(current_event, obj)
 
 
+def _process_payload_step(payload_str, obj):
+    """
+
+    :param encoded:
+    :param obj:
+    :param lambda_context:
```
Comment (aaronkavlie-wf) [inline aws_lambda_fsm/handler.py]: These don't match the function args.
Remove, or correct and document.

File: aws_lambda_fsm/handler.py [hunk 2]
```diff
@@ -75,6 +100,18 @@ def lambda_api_handler(lambda_event):
         logger.exception('Critical error handling lambda: %s', lambda_event)
 
 
+def lambda_step_handler(lambda_event):
+    """
+    AWS Lambda handler for executing state machines.
+
+    :param lambda_event:
+    :return:
```
Comment (aaronkavlie-wf) [inline aws_lambda_fsm/handler.py]: I think descriptions should be added here, or they should be omitted.

File: tools/yaml_to_json.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import json
+
+# library imports
+
+# application imports
+from aws_lambda_fsm.config import get_current_configuration
+from aws_lambda_fsm.constants import CONFIG
+
+# setup the command line args
+parser = argparse.ArgumentParser(description='Turns a fsm.yaml file into a AWS Step Function JSON definition.')
```
Comment (aaronkavlie-wf) [inline tools/yaml_to_json.py]: "~~a~~ an AWS..."

Comment (aaronkavlie-wf) [inline tools/yaml_to_json.py]: The filename and description here seems to be misleading — it suggestions yaml -> json formation translation.

File: tools/yaml_to_json.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                    "BackoffRate": 2.0
+                }
+            ]
+        }
+
+        if state_dict.get(CONFIG.FINAL):
+            data['States'][state_name]['End'] = True
+            continue
+
+        data['States'][state_name]['Next'] = choices_state_name
```
Comment (kevinfrommelt-wf) [inline tools/yaml_to_json.py]: You only need a choices state if the current state has more than one transition?

File: aws_lambda_fsm/handler.py [hunk 5]
```diff
@@ -75,11 +101,23 @@ def lambda_api_handler(lambda_event):
...
+def lambda_step_handler(lambda_event):
+    """
+    AWS Lambda handler for executing state machines.
+
+    :param lambda_event: a dict event from AWS Lambda
+    :return: a dict event to pass along to AWS Step Functions orchestration
+    """
+    obj = {OBJ.SOURCE: AWS.STEP_FUNCTION}
+    payload = json.dumps(lambda_event)  # Step Function just passes straight though
+    return _process_payload_step(payload, obj)
```
Comment (kevinfrommelt-wf) [inline aws_lambda_fsm/handler.py]: It doesn't look like you're able to return Errors in your states, are you planning on adding that at some point?

**Comment by kevinfrommelt-wf** [comment]
Changes look good.

File: aws_lambda_fsm/handler.py (hunk 5 above)
Comment (shawnrusaw-wf) [inline aws_lambda_fsm/handler.py]: i'm not sure what you mean?

File: tools/yaml_to_json.py (hunk 4 above)
Comment (shawnrusaw-wf) [inline tools/yaml_to_json.py]: i think you are right. i'll look to see if i can get that working in a subsequent pr.

File: aws_lambda_fsm/handler.py (hunk 5 above)
Comment (kevinfrommelt-wf) [inline aws_lambda_fsm/handler.py]: It looked like step functions allow you to fail a step with different types of errors that you could then have specific retry / state changes based on the error that was returned.

**Comment by kevinfrommelt-wf** [comment]
+1

**Comment by teresarevious-wf** [comment]
RM +1 - Verified dependencies manually
